File: src/iir/iir_single_sample/iir_single_sample.py
import numpy as np


class IIRSingleSample:
    def __init__(self, b: np.ndarray, a: np.ndarray, filter_order: int):
        """
        Initialize IIR filter.

        Args:
            b (np.ndarray): Numerator coefficients
            a (np.ndarray): Denominator coefficients
            filter_order (int): Filter order 

        Returns:
            None

        Raises:
            ValueError: If the length of the numerator coefficients (b) is not equal to the filter order + 1
            ValueError: If the length of the denominator coefficients (a) is not equal to the filter order + 1
            ValueError: If the length of the numerator and denominator coefficients (b and a respectively) are not equal

        """
        self.b = b
        self.a = a

        if len(b) != filter_order + 1:
            raise ValueError("The length of the numerator coefficients (b) must be equal to the filter order + 1")
        if len(a) != filter_order + 1:
            raise ValueError("The length of the denominator coefficients (a) must be equal to the filter order + 1")
        if len(b) != len(a):
            raise ValueError("The length of the numerator and denominator coefficients (b and a respectively) must be equal")
        
        self.num_taps = filter_order + 1
        self.input_buffer = np.zeros(filter_order)
        self.output_buffer = np.zeros(filter_order)

    def apply_iir_filter(self, x: float) -> float:
        """
        Apply IIR filter to input signal.

        Args:
            x (float): Input sample

        Returns:
            float: Filtered sample
        """
        y = self.b[0] * x
        for i in range(1, len(self.b)):
            y = y + self.b[i] * self.input_buffer[i - 1]
        for i in range(1, len(self.a)):
            y = y - self.a[i] * self.output_buffer[i - 1]
        
        # Rotate the input buffer to the right of one position
        for i in range(1, len(self.b) - 1):
            self.input_buffer[self.num_taps - 1 - i] = self.input_buffer[self.num_taps - 2 - i]
        self.input_buffer[0] = x

        # Rotate the output buffer to the right of one position
        for i in range(1, len(self.a) - 1):
            self.output_buffer[self.num_taps - 1 - i] = self.output_buffer[self.num_taps - 2 -i]
        self.output_buffer[0] = y
        return y

File: src/iir/iir_single_sample/c/iir_single_sample.c
#include "iir_single_sample.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int iir_filter_init(IIRFilter* filter, const double* b, const double* a, int filter_order) {
    filter->num_taps = filter_order + 1;
    
    // Allocate memory for coefficients and buffers
    filter->b = (double*)malloc(filter->num_taps * sizeof(double));
    filter->a = (double*)malloc(filter->num_taps * sizeof(double));
    filter->input_buffer = (double*)malloc(filter_order * sizeof(double));
    filter->output_buffer = (double*)malloc(filter_order * sizeof(double));
    
    if (!filter->b || !filter->a || !filter->input_buffer || !filter->output_buffer) {
        iir_filter_destroy(filter);
        return -1;
    }
    
    // Copy coefficients
    memcpy(filter->b, b, filter->num_taps * sizeof(double));
    memcpy(filter->a, a, filter->num_taps * sizeof(double));
    
    // Initialize buffers to zero
    memset(filter->input_buffer, 0, filter_order * sizeof(double));
    memset(filter->output_buffer, 0, filter_order * sizeof(double));
    
    return 0;
}

double iir_filter_apply(IIRFilter* filter, double x) {
    double y = filter->b[0] * x;
    
    // Apply filter
    for (int i = 1; i < filter->num_taps; i++) {
        y += filter->b[i] * filter->input_buffer[i - 1];
    }
    for (int i = 1; i < filter->num_taps; i++) {
        y -= filter->a[i] * filter->output_buffer[i - 1];
    }
    
    // Rotate input buffer
    for (int i = filter->num_taps - 2; i > 0; i--) {
        filter->input_buffer[i] = filter->input_buffer[i - 1];
    }
    filter->input_buffer[0] = x;
    
    // Print the input buffer
    // printf("Input buffer: ");
    // for (int i = 0; i < filter->num_taps - 1; i++) {
    //     printf("%f ", filter->input_buffer[i]);
    // }
    // printf("\n");

    // Rotate output buffer
    for (int i = filter->num_taps - 2; i > 0; i--) {
        filter->output_buffer[i] = filter->output_buffer[i - 1];
    }
    filter->output_buffer[0] = y;
    
    // Print the output buffer
    // printf("Output buffer: ");
    // for (int i = 0; i < filter->num_taps - 1; i++) {
    //     printf("%f ", filter->output_buffer[i]);
    // }
    // printf("\n");
    
    return y;
}

void iir_filter_destroy(IIRFilter* filter) {
    free(filter->b);
    free(filter->a);
    free(filter->input_buffer);
    free(filter->output_buffer);
}